packages/connect/src/webrtc/connection.ts

```diff
@@ -243,159 +287,157 @@ class WebRTCConnection implements MultiaddrConnection {
+                  case ConnectionEventTypes.PAYLOAD:
+                    if (result.value.done) {
+                      this.flow(`FLOW: webrtc sink: received.done, break`)
+                      reasonToLeave = { type: ConnectionEventTypes.STREAM_ENDED }
+                      leave = true
+                      break
+                    }
+                    toYield = Uint8Array.from([MigrationStatus.NOT_DONE, ...result.value.value])
+                    nextMessage()
+                    break
+                  default:
+                    throw Error(`Invalid result ${JSON.stringify(result)}`)
                 }
 
-                const received = result as StreamResult
-
-                if (received.done) {
-                  this.flow(`FLOW: webrtc sink: received.done, break`)
-                  break
+                if (toYield != undefined) {
+                  this.log(`sinking ${toYield.length} bytes into relayed connection`)
+                  yield toYield
                 }
-
-                next()
-
-                this.log(`sinking ${received.value.slice().length} bytes into relayed connection`)
-                this.flow(`FLOW: webrtc sink: loop iteration ended`)
-                yield Uint8Array.from([MigrationStatus.NOT_DONE, ...received.value.slice()])
               }
               this.flow(`FLOW: webrtc sink: loop ended`)
-              resolve()
+              resolve(reasonToLeave as MigrationEvent | StreamEndedEvent)
             }.call(this)
           )
         )
-        // catch stream errors and forward error
+        // catch stream errors and forward them
         .catch(reject)
     )
 
-    // Either stream is finished or WebRTC is available
-    if (this._webRTCAvailable) {
-      // WebRTC is available, let's attach sink source to it
-      this.flow(`FLOW: sending UPGRADED to relay`)
-      this.relayConn.sendUpgraded()
-
-      // WebRTC handshake was successful, now using direct connection
-      this._sinkMigrated = true
-      if (this._sourceMigrated) {
-        // Update state object once source *and* sink are migrated
-        this.conn = this.relayConn.getWebRTCInstance()
-      }
+    switch (result.type) {
+      case ConnectionEventTypes.STREAM_ENDED:
+        // nothing to do
+        return
+      case ConnectionEventTypes.MIGRATED:
+        // WebRTC is available, let's attach sink source to it
+        this.flow(`FLOW: sending UPGRADED to relay`)
+        this.relayConn.sendUpgraded()
+
+        // WebRTC handshaerrorke was successful, now using direct connection
+        this._sinkMigrated = true
+        if (this._sourceMigrated) {
+          // Update state object once source *and* sink are migrated
+          this.conn = this.relayConn.getWebRTCInstance()
+        }
+        await toIterable.sink(this.relayConn.getWebRTCInstance())(
+          async function* (this: WebRTCConnection): StreamSource {
+            let result: PayloadEvent | SinkSourceAttachedEvent
```

Comment:
Please renable so it does not clash with the `result` in the upper scope. Makes reading easier.

Reply:
changed, renamed both variables